client/service/tofu.go

```diff
@@ -0,0 +1,184 @@
+	slog.Debug("Attempting connection with trusted fingerprint", "fingerprint", fingerprint)
+	dial := func(_ context.Context, network, addr string) (net.Conn, error) {
+		config := &tls.Config{
+			InsecureSkipVerify: true, // We'll verify ourselves
+		}
+
+		conn, err := tls.Dial(network, addr, config)
+		if err != nil {
+			slog.Error("Unable to connect", "addr", addr, "err", err)
+			return nil, err
+
+		}
+		state := conn.ConnectionState()
+		now := time.Now()
+		matched := false
+		for _, cert := range state.PeerCertificates {
+			if fingerprint == getFingerprint(cert.Raw) {
+				matched = true
+			}
+			if now.Before(cert.NotBefore) {
+				_ = conn.Close()
+				return nil, ErrCertNotYetValid
+			}
+			if now.After(cert.NotAfter) {
+				_ = conn.Close()
+				return nil, ErrCertExpired
+			}
+		}
+		if !matched {
+			_ = conn.Close()
+			return nil, ErrNoMatchingFingerprint
+		}
+		// If we've gotten this far, then we can trust the server
+		slog.Debug("Server cert(s) passed TOFU tests")
+		return conn, nil
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			DialTLSContext: dial,
+		},
+	}, nil
+}
+
+// getServerFingerprints connects to an untrusted server, and get cert information about it
+// the addr must be a valid address, including port
+func getServerFingerprints(addr string) ([]CertDetail, error) {
+	var ret []CertDetail
+
+	// If the port isn't specified, assume https standard port
+	if !strings.Contains(addr, ":") {
+		addr = addr + ":443"
+	}
+	slog.Debug("Performing TOFU check", "addr", addr)
+
+	config := &tls.Config{
+		InsecureSkipVerify: true, // By definition, we're trying to build the trust...
+	}
+	conn, err := tls.Dial("tcp", addr, config)
+	if err != nil {
+		slog.Error("Unable to connect to", "addr", addr, "err", err)
+		return nil, err
+	} else {
+		state := conn.ConnectionState()
+
+		for _, cert := range state.PeerCertificates {
+			ret = append(ret, CertDetail{
+				Subject:     cert.Subject.CommonName,
+				Issuer:      cert.Issuer.CommonName,
+				Fingerprint: getFingerprint(cert.Raw),
+			})
+		}
+		if len(ret) == 0 {
+			return nil, ErrNoCertsDetected
+		}
+		return ret, nil
+	}
+}
+
+// TrustFingerprintCallback is a callback function that is called when a server's certificate
+// fingerprint is detected. It receives the server's IP address and the certificate details.
+// The responder can return a CertDetail to be used as the trusted fingerprint.
+// If the returned CertDetail is nil, the connection will be rejected.
+type TrustFingerprintCallback func(ip string, details []CertDetail) *CertDetail
+
+func readTrustedServerFingerprints(fingerprintsJSON string) (map[string]string, error) {
+	fingerprints := map[string]string{}
+	if _, err := os.Stat(fingerprintsJSON); err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return fingerprints, nil
+		}
+		return nil, fmt.Errorf("failed to check trusted server fingerprints: %w", err)
+	}
+
+	data, err := os.ReadFile(fingerprintsJSON)
```

Comment:
You can collapse the `os.Stat()` call into the `os.ReadFile()` call and avoid a potential [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) scenario.